### PR TITLE
fix: get context back for composite spans

### PR DESCRIPTION
**Pull request by astorm**
Closes: https://github.com/elastic/apm-agent-nodejs/issues/2715

When span compression is enabled and a span is compression eligible ensure the context is not nulled out in order to ensure its available for span serialization/encoding. 

### Checklist

- [x] Implement code
- [x] Add tests
~- [ ] Update TypeScript typings~
~- [ ] Update documentation~
- [x] Add CHANGELOG.asciidoc entry
~- [ ] Commit message follows [commit guidelines](https://github.com/elastic/apm-agent-nodejs/blob/main/CONTRIBUTING.md#commit-message-guidelines)~

### CODE REVIEW

File: lib/instrumentation/span.js [hunk 1]
```diff
@@ -119,16 +119,20 @@ Span.prototype.end = function (endTime) {
...
-  this._http = null
-  this._message = null
-  this._capturedStackTrace = null
-  // GenericSpan fields:
-  // - Cannot drop `this._context` because it is used for traceparent and ids.
-  this._timer = null
-  this._labels = null
+  // required to support `interface Span`. References can't be dropped if this
+  // is a potential composite span since a composite span will encode _after_
+  // it has ended.
```
[the remark this answers is not in the document]
Comment (astorm) [reply lib/instrumentation/span.js]: I moved the reference block to the `done` callback of the encode method to ensure the the span payload creation in the main body of `Span.prototype._encode` will have access to there methods regardless of where/when a span is serialized/encoded and that we can still take advantage of the dereferencing regardless of compression setting.   

GHA tests seem happy and I presumes jenkins will grind its way to green, resolving.

File: test/instrumentation/github-issue-2715.test.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  centralConfig: false,
+  cloudProvider: 'none',
+  spanCompressionEnabled: true
+})
+
+const mockClient = require('../_mock_http_client')
+
+const tape = require('tape')
+tape.test(function (suite) {
+  suite.test(function (t) {
```
Comment (trentm) [inline test/instrumentation/github-issue-2715.test.js:14]: nits:
- Why the wrap in a suite here for a single test?
- Both the `tape.test(` and `suite.test(` should get a test name to assist with self-documenting and correlation with output. Currently it looks like:

```
% node test/instrumentation/github-issue-2715.test.js
TAP version 13
# (anonymous)
# (anonymous)
ok 1 should be strictly equal
ok 2 should be strictly equal
ok 3 should be strictly equal
ok 4 should be strictly equal


{"log.level":"error","@timestamp":"2022-05-24T17:58:52.688Z","log":{"logger":"elastic-apm-node"},"ecs":{"version":"1.6.0"},"message":"APM Server transport error: error fetching APM Server version: connect ECONNREFUSED 127.0.0.1:8200"}

1..4
# tests 4
# pass  4
```

File: test/instrumentation/github-issue-2715.test.js [hunk 3]
```diff
@@ -0,0 +1,52 @@
+const agent = require('../..').start({
+  serviceName: 'test-composite-context',
+  captureExceptions: false,
+  metricsInterval: 0,
+  centralConfig: false,
+  cloudProvider: 'none',
+  spanCompressionEnabled: true
```
Comment (trentm) [inline test/instrumentation/github-issue-2715.test.js:7]: nit: The `APM Server transport error: error fetching APM Server version:` test output error (shown below) could be avoided by setting `apmServerVersion: '8.0.0',` or similar in the agent config here.

**Review by trentm — APPROVED** [review]
Just a couple nits in the test file.

**Pull request merged**